### PR TITLE
[WIP, DO NOT MERGE] Fix conflicting port and lifecycle management for distributed schedulers

### PR DESCRIPTION
**Pull request by zhreshold**
*Issue #, if available:*
Attempts to improve the scheduler stability on distributed nodes. There are known failure cases caused by `port is already allocated` before remote scheduler and worker to exit disgracefully

*Description of changes:*

- Remove global lock based PORT_ID assignment to remote schedulers
- The scheduler will still be launched in remote nodes due to the constraints in scheduler reporter synchronization, however, instead of using fixed port_id(potential conflict), it now support automatic port finder with dask's builtin `SSHCluster`
- Fixed a dask issue that prevent `no-nanny` worker to be launched with standard cluster setting. 
- Removed usage of manually managed background SSH connections

**note**
code for port_id, ssh_helper may not be necessary anymore after this PR, but I am waiting for more testings before being confidently removing non critical code removals.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by zhreshold** [comment]
@tlienart @Innixma @jwmueller @sxjscience please comment

**Comment by tlienart** [comment]
This is great, I won't be able to launch experiments with this before next week but will do so and report. Thanks!

Tagging @aaronkl as well

**Comment by szha** [comment]
Job PR-649-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-649/1/index.html

**Comment by szha** [comment]
Job PR-649-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-649/2/index.html

**Comment by mseeger** [comment]
@tlienart Any news on this here?

File: setup.py [hunk 1]
```diff
@@ -59,7 +59,8 @@ def create_version_file():
     'pandas>=0.24.0,<1.0',
     'psutil>=5.0.0,<=5.7.0',  # TODO: psutil 5.7.1/5.7.2 has non-deterministic error on CI doc build -  ImportError: cannot import name '_psutil_linux' from 'psutil'
     'scikit-learn>=0.22.0,<0.23',
-    'networkx>=2.3,<3.0'
+    'networkx>=2.3,<3.0',
+    'asyncssh'
```
Comment (Innixma) [inline setup.py:63]: why is asyncssh required?

**Comment by tlienart** [comment]
I believe @aaronkl ran experiments with this recently and is more up to date on this? Not sure whether he tested this PR explicitly, I haven't had the occasion to test it but could if required

File: setup.py (hunk 1 above)
Comment (zhreshold) [reply setup.py:63]: it's required by SSHCluster in dask

**Comment by zhreshold** [comment]
@tlienart Sorry I haven't had a chance to update this PR recently but it will be great if you or Aaron can try it with internal setup.

**Comment by tlienart** [comment]
I tested this today, here's the stack trace some may be due to me having made a mistake but it looks like new errors to me:

```
Traceback (most recent call last):
  File "main.py", line 26, in <module>
    train(params)
  File "/opt/ml/code/train.py", line 99, in train
    searcher_data=options['searcher_data'])
  File "/autogluon-replica/autogluon/scheduler/hyperband.py", line 203, in __init__
    delay_get_config=delay_get_config)
  File "/autogluon-replica/autogluon/scheduler/fifo.py", line 116, in __init__
    super().__init__(dist_ip_addrs)
  File "/autogluon-replica/autogluon/scheduler/scheduler.py", line 30, in __init__
    remotes = cls.REMOTE_MANAGER.add_remote_nodes(dist_ip_addrs)
  File "/autogluon-replica/autogluon/scheduler/remote/remote_manager.py", line 78, in add_remote_nodes
    remote = Remote(node_ip, port)
  File "/autogluon-replica/autogluon/scheduler/remote/remote.py", line 76, in __init__
    scheduler_options=
{
    "port": 0,
    "dashboard_address": ":8787"
}
)
  File "/usr/local/lib/python3.6/site-packages/distributed/deploy/ssh.py", line 371, in SSHCluster
    return SpecCluster(workers, scheduler, name="SSHCluster", **kwargs)
  File "/usr/local/lib/python3.6/site-packages/distributed/deploy/spec.py", line 276, in __init__
    self.sync(self._start)
  File "/usr/local/lib/python3.6/site-packages/distributed/deploy/cluster.py", line 174, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/distributed/utils.py", line 339, in sync
    raise exc.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/distributed/utils.py", line 323, in f
    result[0] = yield future
  File "/usr/local/lib/python3.6/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/usr/local/lib/python3.6/site-packages/distributed/deploy/spec.py", line 304, in _start
    self.scheduler = await self.scheduler
  File "/usr/local/lib/python3.6/site-packages/distributed/deploy/spec.py", line 71, in _
    await self.start()
  File "/usr/local/lib/python3.6/site-packages/distributed/deploy/ssh.py", line 174, in start
    self.connection = await asyncssh.connect(self.address, **self.connect_options)
  File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 6416, in connect
    conn_factory, 'Opening SSH connection to')
  File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 218, in _connect
    await conn.wait_established()
  File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 2096, in wait_established
    await self._waiter
  File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 910, in data_received
    while self._inpbuf and self._recv_handler():
  File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 1143, in _recv_packet
    processed = handler.process_packet(pkttype, seq, packet)
  File "/usr/local/lib/python3.6/site-packages/asyncssh/packet.py", line 215, in process_packet
    self._packet_handlers[pkttype](self, pkttype, pktid, packet)
  File "/usr/local/lib/python3.6/site-packages/asyncssh/kex_dh.py", line 252, in _process_reply
    host_key = self._conn.validate_server_host_key(host_key_data)
  File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 2698, in validate_server_host_key
    raise HostKeyNotVerifiable(str(exc)) from None
Traceback (most recent call last): File "main.py", line 26, in <module> train(params) File "/opt/ml/code/train.py", line 99, in train searcher_data=options['searcher_data']) File "/autogluon-replica/autogluon/scheduler/hyperband.py", line 203, in __init__ delay_get_config=delay_get_config) File "/autogluon-replica/autogluon/scheduler/fifo.py", line 116, in __init__ super().__init__(dist_ip_addrs) File "/autogluon-replica/autogluon/scheduler/scheduler.py", line 30, in __init__ remotes = cls.REMOTE_MANAGER.add_remote_nodes(dist_ip_addrs) File "/autogluon-replica/autogluon/scheduler/remote/remote_manager.py", line 78, in add_remote_nodes remote = Remote(node_ip, port) File "/autogluon-replica/autogluon/scheduler/remote/remote.py", line 76, in __init__ scheduler_options={"port": 0, "dashboard_address": ":8787"}) File "/usr/local/lib/python3.6/site-packages/distributed/deploy/ssh.py", line 371, in SSHCluster return SpecCluster(workers, scheduler, name="SSHCluster", **kwargs) File "/usr/local/lib/python3.6/site-packages/distributed/deploy/spec.py", line 276, in __init__ self.sync(self._start) File "/usr/local/lib/python3.6/site-packages/distributed/deploy/cluster.py", line 174, in sync return sync(self.loop, func, *args, **kwargs) File "/usr/local/lib/python3.6/site-packages/distributed/utils.py", line 339, in sync raise exc.with_traceback(tb) File "/usr/local/lib/python3.6/site-packages/distributed/utils.py", line 323, in f result[0] = yield future File "/usr/local/lib/python3.6/site-packages/tornado/gen.py", line 735, in run value = future.result() File "/usr/local/lib/python3.6/site-packages/distributed/deploy/spec.py", line 304, in _start self.scheduler = await self.scheduler File "/usr/local/lib/python3.6/site-packages/distributed/deploy/spec.py", line 71, in _ await self.start() File "/usr/local/lib/python3.6/site-packages/distributed/deploy/ssh.py", line 174, in start self.connection = await asyncssh.connect(self.address, **self.connect_options) File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 6416, in connect conn_factory, 'Opening SSH connection to') File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 218, in _connect await conn.wait_established() File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 2096, in wait_established await self._waiter File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 910, in data_received while self._inpbuf and self._recv_handler(): File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 1143, in _recv_packet processed = handler.process_packet(pkttype, seq, packet) File "/usr/local/lib/python3.6/site-packages/asyncssh/packet.py", line 215, in process_packet self._packet_handlers[pkttype](self, pkttype, pktid, packet) File "/usr/local/lib/python3.6/site-packages/asyncssh/kex_dh.py", line 252, in _process_reply host_key = self._conn.validate_server_host_key(host_key_data) File "/usr/local/lib/python3.6/site-packages/asyncssh/connection.py", line 2698, in validate_server_host_key raise HostKeyNotVerifiable(str(exc)) from None

2020-09-22T14:47:10.694+02:00

Copy
asyncssh.misc.HostKeyNotVerifiable: Host key is not trusted
asyncssh.misc.HostKeyNotVerifiable: Host key is not trusted
```

further

```
Exception ignored in: <object repr() failed>
Exception ignored in: <object repr() failed>

2020-09-22T14:47:12.695+02:00	Traceback (most recent call last): File "/usr/local/lib/python3.6/site-packages/distributed/client.py", line 1192, in __del__ File "/autogluon-replica/autogluon/scheduler/remote/remote.py", line 85, in close File "/usr/local/lib/python3.6/site-packages/distributed/client.py", line 1404, in close

AttributeError: 'Remote' object has no attribute 'status'
AttributeError: 'Remote' object has no attribute 'status'
```

```
2020-09-22 12:47:12,225 sagemaker-training-toolkit ERROR    ExecuteUserScriptError:
2020-09-22 12:47:12,225 sagemaker-training-toolkit ERROR ExecuteUserScriptError:
```

**Comment by zhreshold** [comment]
not sure how asyncssh examines the known_hosts which might need a bit polishment in sagemaker setup

**Pull request closed without merging**